### ISSUE TITLE
Implementation details

### ISSUE THREAD

**Issue by rayleizhu** [SHI-Labs/NATTEN#27]
I'm writing my own CUDA extension using your awsome code as the template. Here are some questions about the implementation details:

(1) Why do you use torch::DefaultPtrTraits instead of torch::RestrictPtrTraits as recommended by [pytorch official extension guide](https://pytorch.org/tutorials/advanced/cpp_extension.html)? Is it for performance consideration?

https://github.com/SHI-Labs/NATTEN/blob/a456a470d485e6e3b74bc598fa3f048d53414730/src/natten/csrc/cuda/neighborhood_neighborhood_2d.cu#L44

(2) Why do you calculate offsets manually, instead of using indexing operator `[]` as done in official guide ?

https://github.com/SHI-Labs/NATTEN/blob/a456a470d485e6e3b74bc598fa3f048d53414730/src/natten/csrc/cuda/neighborhood_neighborhood_2d.cu#L75

**Comment by alihassanijr** [comment]
Hello and thank you for your interest.

1. I believe the difference is that RestrictPtrTraits just prepends the `__restrict__` tag to the data pointers. Not sure why or when we went from RestrictPtrTraits (that is if we didn't start off with DefaultPtrTraits), but I don't think it makes that big of a difference unless one sets up a different caching policy which might be more useful for the Q K V tensors than anything else.
But thanks for bringing it up, we might run some tests to see how RestrictPtrTraits plays out.

2. If I recall correctly the switch was made when we started doing fp16 kernels, because referencing by brackets on `half` addresses was a little tricky, so we switched entirely to addresses and offsets. Not sure if it's more optimal, but I doubt if it's less optimal, especially since certain (partial) offsets can be computed just once and re-applied to similar tensors.
